sdks/go/test/run_validatesrunner_tests.sh

```diff
@@ -0,0 +1,174 @@
+# services can be built from the repository.
+
+set -e
+set -v
+
+RUNNER=portable
+
+exit_background_processes () {
+  if [[ -n "$JOBSERVER_PID" ]]; then
+    kill -s SIGKILL $JOBSERVER_PID
+  fi
+  if [[ -n "$EXPANSION_PID" ]]; then
+    kill -s SIGKILL $EXPANSION_PID
+  fi
+}
+trap exit_background_processes SIGINT SIGTERM EXIT
+
+while [[ $# -gt 0 ]]
+do
+key="$1"
+case $key in
+    --runner)
+        RUNNER="$2"
+        shift # past argument
+        shift # past value
+        ;;
+    --flink_job_server_jar)
+        FLINK_JOB_SERVER_JAR="$2"
+        shift # past argument
+        shift # past value
+        ;;
+    --spark_job_server_jar)
+        SPARK_JOB_SERVER_JAR="$2"
+        shift # past argument
+        shift # past value
+        ;;
+    --endpoint)
+        ENDPOINT="$2"
+        shift # past argument
+        shift # past value
+        ;;
+    --expansion_service_jar)
+        EXPANSION_SERVICE_JAR="$2"
+        shift # past argument
+        shift # past value
+        ;;
+    --expansion_addr)
+        EXPANSION_ADDR="$2"
+        shift # past argument
+        shift # past value
+        ;;
+    *)    # unknown option
+        echo "Unknown option: $1"
+        exit 1
+        ;;
+esac
+done
+
+# Go to the root of the repository
+cd $(git rev-parse --show-toplevel)
+
+# Verify in the root of the repository
+test -d sdks/go/test
+
+# Set up environment based on runner.
+ARGS=--runner=$RUNNER
+if [[ "$RUNNER" == "flink" || "$RUNNER" == "spark" || "$RUNNER" == "portable" ]]; then
+  if [[ -z "$ENDPOINT" ]]; then
+    # Hacky python script to find a free port. Note there is a small chance the chosen port could
+    # get taken before being claimed by the job server.
+    SOCKET_SCRIPT="
+import socket
+s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+s.bind(('localhost', 0))
+print(s.getsockname()[1])
+s.close()
+    "
+    JOB_PORT=$(python -c "$SOCKET_SCRIPT")
+    ENDPOINT="localhost:$JOB_PORT"
+    echo "No endpoint specified; starting a new $RUNNER job server on $ENDPOINT"
+    if [[ "$RUNNER" == "flink" ]]; then
+      java \
+          -jar $FLINK_JOB_SERVER_JAR \
+          --flink-master [local] \
+          --job-port $JOB_PORT \
+          --expansion-port 0 \
+          --artifact-port 0 &
+    elif [[ "$RUNNER" == "spark" ]]; then
+      java \
+          -jar $SPARK_JOB_SERVER_JAR \
+          --spark-master-url local \
+          --job-port $JOB_PORT \
+          --expansion-port 0 \
+          --artifact-port 0 &
+    elif [[ "$RUNNER" == "portable" ]]; then
+      python \
+          -m apache_beam.runners.portability.local_job_service_main \
+          --port $JOB_PORT &
+    else
+      echo "Unknown runner: $RUNNER"
+      exit 1;
+    fi
+    JOBSERVER_PID=$!
+  fi
+
+  if [[ -z "$EXPANSION_ADDR" ]]; then
+    SOCKET_SCRIPT="
+import socket
+s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+s.bind(('localhost', 0))
+print(s.getsockname()[1])
+s.close()
+    "
```

Comment:
Deduplicate w/the socket script higher up please.

Reply:
Done.